gazebo/dave_gz_sensor_plugins/src/sea_pressure_sensor.cc

```diff
@@ -0,0 +1,291 @@
+          << "plugin will not be initialized." << std::endl;
+    return;
+  }
+  this->dataPtr->robotNamespace = _sdf->Get<std::string>("namespace");
+
+  if (_sdf->HasElement("saturation"))
+  {
+    this->dataPtr->saturation = _sdf->Get<double>("saturation");
+  }
+  else
+  {
+    this->dataPtr->saturation = 3000;  // original xacro file has it as range set at 30000
+  }
+
+  if (_sdf->HasElement("estimate_depth_on"))
+  {
+    this->dataPtr->estimateDepth = _sdf->Get<bool>("estimate_depth_on");
+  }
+  else
+  {
+    this->dataPtr->estimateDepth = false;
+  }
+
+  if (_sdf->HasElement("standard_pressure"))
+  {
+    this->dataPtr->standardPressure = _sdf->Get<double>("standard_pressure");
+  }
+  else
+  {
+    this->dataPtr->standardPressure = 101.325;
+  }
+
+  if (_sdf->HasElement("kPa_per_meter"))
+  {
+    this->dataPtr->kPaPerM = _sdf->Get<double>("kPa_per_meter");
+  }
+  else
+  {
+    this->dataPtr->kPaPerM = 9.80638;
+  }
+
+  // this->dataPtr->gazeboNode->Init();
+  this->dataPtr->modelEntity = GetModelEntity(this->dataPtr->robotNamespace, _ecm);  // check TODO
+
+  this->dataPtr->rosSensorOutputPub =
+    this->rosNode->create_publisher<sensor_msgs::msg::FluidPressure>(
+      "sensor_output_topic", rclcpp::QoS(10).reliable());
+
+  if (this->dataPtr->gzMsgEnabled)
+  {
+    this->dataPtr->gazeboSensorOutputPub =
+      this->dataPtr->gazeboNode->Advertise<pressure_sensor_msgs::msgs::Pressure>(
+        this->dataPtr->robotNamespace + "/" + "sensor_output_topic");
+  }
+}
+//////////////////////////////////////////
+
+gz::sim::Entity SubseaPressureSensorPlugin::GetModelEntity(
+  const std::string & modelName, gz::sim::EntityComponentManager & ecm)
+{
+  gz::sim::Entity modelEntity = gz::sim::kNullEntity;
+
+  ecm.Each<gz::sim::components::Name>(
+    [&](const gz::sim::Entity & entity, const gz::sim::components::Name * nameComp) -> bool
+    {
+      if (nameComp->Data() == modelName)
+      {
+        modelEntity = entity;
+        return false;  // Stop iteration
+      }
+      return true;  // Continue iteration
+    });
+
+  return modelEntity;
+}
+
+gz::math::Pose3d SubseaPressureSensorPlugin::GetModelPose(
+  const gz::sim::Entity & modelEntity, gz::sim::EntityComponentManager & ecm)
+{
+  const auto * poseComp = ecm.Component<gz::sim::components::Pose>(modelEntity);
+  if (poseComp)
+  {
+    return poseComp->Data();
+  }
+  else
+  {
+    gzerr << "Pose component not found for entity: " << modelEntity << std::endl;
+    return gz::math::Pose3d::Zero;
+  }
+}
+
+void SubseaPressureSensorPlugin::PreUpdate(
+  const gz::sim::UpdateInfo & _info, gz::sim::EntityComponentManager & _ecm)
+{
+  // Get model pose
+
+  gz::math::Pose3d sea_pressure_sensor_pos = GetModelPose(this->dataPtr->modelEntity, _ecm);
+
+  // const gz::sim::Entity & _entity = _ecm.EntityByComponents(gz::sim::components::WorldPose());
+  // gz::math::Vector3 pos;
+  // auto pos = GetWorldPose(_entity, _ecm);
```

Comment:
Please remove any of the unnecessary codes. Many other exists in this file.

Reply:
Done !!